gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java

```diff
@@ -451,13 +452,26 @@ private Task createTaskRunnable(WorkUnitState workUnitState, CountDownLatch coun
 
   public void runAndOptionallyCommitTaskAttempt(CommitPolicy multiTaskAttemptCommitPolicy)
       throws IOException, InterruptedException {
-    run();
-    if (multiTaskAttemptCommitPolicy.equals(GobblinMultiTaskAttempt.CommitPolicy.IMMEDIATE)) {
-      this.log.info("Will commit tasks directly.");
-      commit();
-    } else if (!isSpeculativeExecutionSafe()) {
-      throw new RuntimeException(
-          "Speculative execution is enabled. However, the task context is not safe for speculative execution.");
+    try {
+      run();
+      if (multiTaskAttemptCommitPolicy.equals(GobblinMultiTaskAttempt.CommitPolicy.IMMEDIATE)) {
+        this.log.info("Will commit tasks directly.");
+        commit();
+      } else if (!isSpeculativeExecutionSafe()) {
+        throw new RuntimeException(
+            "Speculative execution is enabled. However, the task context is not safe for speculative execution.");
+      }
+    } finally {
+      // During the task execution, the fork/task instances will create metric contexts (fork, task, job, container)
+      // along the hierarchy up to the root metric context. Although root metric context has a weak reference to
+      // those metric contexts, they are meanwhile cached by GobblinMetricsRegistry. Here we will remove all those
+      // strong reference from the cache to make sure it can be reclaimed by Java GC when JVM has run out of memory.
+
+      this.tasks.forEach(task-> {
+        TaskMetrics.remove(task);
+      });
+
+      JobMetrics.remove(GobblinMetrics.METRICS_ID_PREFIX + jobState.getJobId());
```

Comment:
I don't think this should be done in here, the job will still run after the tasks are done. If this is needed for distributed mode, then let's check that we are running on a worker before removing the job metrics.

Reply:
Most of our Gobblin code runs in distributed mode, like MRJobLauncher, HelixJobLauncher, etc. I will add the defensive check to make sure that this GobblinMultiTaskAttempt was actually in the separate process.

But even for LocalJobLauncher, GobblinMetricsRegistry is actually a cache, there is no guarantee that JobMetrics cannot be evicted if TaskMetrics is still alive.

Reply:
@ibuenros Please check my latest change where a creatorTag was added to make sure the JobMetrics can be removed by the correct owner.